File: 04_BehaviouralDesignPatterns/09_MementoPattern/java/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

// memento pattern implementation for a resume editor
class ResumeEditor {
    private String name;
    private String education;
    private String experience;
    private List<String> skills;

    public void setName(String name) {
        this.name = name;
    }

    public void setEducation(String education) {
        this.education = education;
    }

    public void setExperience(String experience) {
        this.experience = experience;
    }

    public void setSkills(List<String> skills) {
        this.skills = new ArrayList<>(skills);
    }

    public void printResume() {
        System.out.println("----- Resume -----");
        System.out.println("Name: " + name);
        System.out.println("Education: " + education);
        System.out.println("Experience: " + experience);
        System.out.println("Skills: " + String.join(", ", skills));
        System.out.println("----- Resume -----");
    }

    public Memento save() {
        return new Memento(name, education, experience, List.copyOf(skills));
    }

    public void restore(Memento memento) {
        this.name = memento.name;
        this.education = memento.education;
        this.experience = memento.experience;
        this.skills = new ArrayList<>(memento.skills);
    }

    // Memento class to hold the state of the ResumeEditor
    public static class Memento {
        private final String name;
        private final String education;
        private final String experience;
        private final List<String> skills;

        private Memento(String name, String education, String experience, List<String> skills) {
            this.name = name;
            this.education = education;
            this.experience = experience;
            this.skills = new ArrayList<>(skills);
        }

        public String getName() {
            return name;
        }

        public String getEducation() {
            return education;
        }

        public String getExperience() {
            return experience;
        }

        public List<String> getSkills() {
            return skills;
        }
    }
}

class ResumeHistory {
    private Stack<ResumeEditor.Memento> history = new Stack<>();

    public void save(ResumeEditor editor) {
        history.push(editor.save());
    }

    public void undo(ResumeEditor editor) {
        if (!history.isEmpty()) {
            editor.restore(history.pop());
        } else {
            System.out.println("No more actions to undo.");
        }
    }
}

public class Main {
    public static void main(String[] args) {
        ResumeEditor editor = new ResumeEditor();
        ResumeHistory history = new ResumeHistory();

        // Initial resume state
        editor.setName("John Doe");
        editor.setEducation("B.Sc. in Computer Science");
        editor.setExperience("5 years in software development");
        editor.setSkills(List.of("Java", "Python", "SQL"));
        editor.printResume();

        // Save the current state
        history.save(editor);

        // Modify the resume
        editor.setExperience("6 years in software development");
        editor.setSkills(List.of("Java", "Python", "SQL", "JavaScript"));
        editor.printResume();

        // Undo the last change
        history.undo(editor);
        editor.printResume();

        // Undo again
        history.undo(editor);
        editor.printResume();
    }
}

File: 04_BehaviouralDesignPatterns/09_MementoPattern/cpp/Main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <stack>
#include <ctime>

// Memento class - stores the state
class ResumeMemento {
private:
    friend class Resume; // Only Resume can access private members

    struct State {
        std::string name;
        std::string education;
        std::string experience;
        std::vector<std::string> skills;
        std::time_t timestamp;
    };

    State state;

    explicit ResumeMemento(const State& state) : state(state) {}

public:
    std::time_t getTimestamp() const {
        return state.timestamp;
    }
};

// Originator class
class Resume {
private:
    std::string name;
    std::string education;
    std::string experience;
    std::vector<std::string> skills;

public:
    // Setters with validation
    void setName(const std::string& name) {
        if (name.empty()) {
            throw std::invalid_argument("Name cannot be empty");
        }
        this->name = name;
    }

    void setEducation(const std::string& education) {
        if (education.empty()) {
            throw std::invalid_argument("Education cannot be empty");
        }
        this->education = education;
    }

    void setExperience(const std::string& experience) {
        if (experience.empty()) {
            throw std::invalid_argument("Experience cannot be empty");
        }
        this->experience = experience;
    }

    void addSkill(const std::string& skill) {
        if (skill.empty()) {
            throw std::invalid_argument("Skill cannot be empty");
        }
        skills.push_back(skill);
    }

    // Create a memento containing a snapshot of current state
    std::unique_ptr<ResumeMemento> save() const {
        ResumeMemento::State state{
            name,
            education,
            experience,
            skills,
            std::time(nullptr)
        };
        return std::make_unique<ResumeMemento>(state);
    }

    // Restore state from a memento
    void restore(const ResumeMemento& memento) {
        name = memento.state.name;
        education = memento.state.education;
        experience = memento.state.experience;
        skills = memento.state.skills;
    }

    // Display current state
    void display() const {
        std::cout << "Name: " << name << std::endl;
        std::cout << "Education: " << education << std::endl;
        std::cout << "Experience: " << experience << std::endl;
        std::cout << "Skills: ";
        for (size_t i = 0; i < skills.size(); ++i) {
            std::cout << skills[i];
            if (i < skills.size() - 1) {
                std::cout << ", ";
            }
        }
        std::cout << std::endl;
    }
};

// Caretaker class - maintains history of mementos
class ResumeHistory {
private:
    Resume& resume;
    std::stack<std::unique_ptr<ResumeMemento>> history;
    std::stack<std::unique_ptr<ResumeMemento>> redoHistory;
    const size_t maxHistory;

public:
    explicit ResumeHistory(Resume& resume, size_t maxHistory = 10) 
        : resume(resume), maxHistory(maxHistory) {}

    void save() {
        // Clear redo history when new change is made
        while (!redoHistory.empty()) {
            redoHistory.pop();
        }

        history.push(resume.save());
        
        // Maintain history size limit
        if (history.size() > maxHistory) {
            history.pop();
        }
    }

    bool undo() {
        if (history.empty()) {
            std::cout << "No more history to undo" << std::endl;
            return false;
        }

        redoHistory.push(resume.save());
        resume.restore(*history.top());
        history.pop();
        return true;
    }

    bool redo() {
        if (redoHistory.empty()) {
            std::cout << "No more changes to redo" << std::endl;
            return false;
        }

        history.push(resume.save());
        resume.restore(*redoHistory.top());
        redoHistory.pop();
        return true;
    }

    void displayHistory() const {
        std::cout << "\nUndo history timestamps:" << std::endl;
        std::stack<std::unique_ptr<ResumeMemento>> temp = history;
        while (!temp.empty()) {
            std::time_t timestamp = temp.top()->getTimestamp();
            std::cout << std::ctime(&timestamp);
            temp.pop();
        }
    }
};

int main() {
    try {
        Resume resume;
        ResumeHistory history(resume);

        // Initial state
        resume.setName("John Doe");
        resume.setEducation("Bachelor of Science in Computer Science");
        resume.setExperience("5 years of software development experience");
        resume.addSkill("Java");
        resume.addSkill("Python");
        history.save();
        std::cout << "Initial Resume:" << std::endl;
        resume.display();

        // Make some changes
        std::cout << "\nUpdating resume..." << std::endl;
        resume.setName("John Smith");
        resume.setEducation("Master of Science in Computer Science");
        resume.setExperience("6 years of software development experience");
        resume.addSkill("JavaScript");
        history.save();
        resume.display();

        // Undo changes
        std::cout << "\nUndo last changes:" << std::endl;
        history.undo();
        resume.display();

        // Redo changes
        std::cout << "\nRedo last changes:" << std::endl;
        history.redo();
        resume.display();

        // Show history
        history.displayHistory();

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}